src/java.base/share/classes/java/util/stream/Collectors.java

```diff
@@ -298,8 +298,14 @@ private static IllegalStateException duplicateKeyException(
     Collector<T, ?, List<T>> toUnmodifiableList() {
         return new CollectorImpl<>((Supplier<List<T>>) ArrayList::new, List::add,
                                    (left, right) -> { left.addAll(right); return left; },
-                                   list -> (List<T>)SharedSecrets.getJavaUtilCollectionAccess()
-                                                                 .listFromTrustedArray(list.toArray()),
+                                   list -> {
+                                       if (list.getClass() == ArrayList.class) { // ensure it's trusted
+                                           return (List<T>)SharedSecrets.getJavaUtilCollectionAccess()
```

Comment:
While we are in this method i think it should be possible to remove the all casts and the `@SuppressWarnings` (javac's type inference got better after this method was added?).

Reply:
I've updated this PR to clean this up. Not sure where this came from in the first place. It might have been left over from an intermediate edit.